EstEID 2018+ driver

martinpaljak opened the pull request:
This adds card driver for Estonian eID from 2018 onwards.

Next iteration will merge pkcs15-esteid and get rid of Estonia related stuff in card-mcrd.c

plaes commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+     "EstEID 2018", SC_CARD_TYPE_ESTEID_2018, 0, NULL},
+    {NULL, NULL, NULL, 0, 0, NULL}};
+
+static struct sc_aid IASECC_AID = {{0xA0, 0x00, 0x00, 0x00, 0x77, 0x01, 0x08,
```
const

plaes commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+     "EstEID 2018", SC_CARD_TYPE_ESTEID_2018, 0, NULL},
+    {NULL, NULL, NULL, 0, 0, NULL}};
+
+static struct sc_aid IASECC_AID = {{0xA0, 0x00, 0x00, 0x00, 0x77, 0x01, 0x08,
```
const

martinpaljak replied:
check

plaes commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+      {
+        file = sc_file_new();
+        if (file == NULL)
+          LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
```
leaking the `path` here?

plaes commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+  struct esteid_priv_data *priv;
+  sc_apdu_t apdu;
+
+  unsigned char cse_crt_aut[] = {0x80, 0x04, 0xFF, 0x20, 0x08,
```
all three `const`?

martinpaljak replied:
check

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,226 @@
+    pin_info.attrs.pin.stored_length = 12;
+    pin_info.attrs.pin.max_length = 12;
+    pin_info.attrs.pin.pad_char = 0xFF;
+    pin_info.tries_left = 3;
```
Shouldn't tries_left be `-1` if it's unknown?

martinpaljak replied:
It gets updated after initial adding and 3 is the specification-defined value.

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,226 @@
+  for (i = 0; i < r; i++)
+  {
+    // trigger PIN counter refresh via pin_cmd
+    r = sc_pkcs15_get_pin_info(p15card, objs[i]);
```
why is this needed? this should be called automatically if the application requests the pin info

martinpaljak replied:
For `pkcs15-tool -D/--list-pins`. Which could of course be changed to try to update that info when not available.

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,226 @@
+    strlcpy(prkey_obj.label, prkey_name[i], sizeof(prkey_obj.label));
+    prkey_obj.auth_id.len = 1;
+    prkey_obj.auth_id.value[0] = prkey_pin[i];
+    prkey_obj.user_consent = (i == 1 ? 1 : 0);
```
use `int prkey_consent[2] = {0, 1};` above

martinpaljak replied:
Done

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+    apdu.datalen = 2;
+  }
+  apdu.le = 0x00;
+  apdu.resplen = 0;
```
Just discovered that, somehow my notifications for this repo were turned off and I based it on 0.19.0 tag. Will update.w

martinpaljak replied:
Which made me think: why not just have a changed short name like "sc_make_apdu" instead of the weird-looking _ex?

frankmorgner replied:
adding _ex lets developers know which one is the successor. Introducing new names adds confusion. Ideally, I'd like to remove the function without the _ex, but I'm too lazy to go through all the codebase...

martinpaljak commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,226 @@
+  {
+    static const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+    static char const *esteid_cert_paths[2] = {"3f00adf13401", "3f00adf2341f"};
+    static int esteid_cert_ids[2] = {1, 2};
```
None, to be removed.

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+        file = sc_file_new();
+        if (file == NULL)
+          LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+        file->path = *in_path;
```
They don't seem to be necessary?

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+static struct sc_aid IASECC_AID = {{0xA0, 0x00, 0x00, 0x00, 0x77, 0x01, 0x08,
+                                    0x00, 0x07, 0x00, 0x00, 0xFE, 0x00, 0x00,
+                                    0x01, 0x00},
+                                   16};
```
Yes and no.
- card-iasecc.c is indeed too bloated and complex already, this complexity is not needed for Estonian card (speed/reliability)
- card-esteid.c (once amended) would also integrate pieces from card-mcrd.c that work with pre-iasecc cards in Estonia. I gave it a long thought and decided that a separate driver is more meaningful than trying to amend or refactor card-iasecc.c

It comes down to a slim card driver that does only what is necessary (working PKCS#11) in a static setup without any metadata reading or parsing.

frankmorgner replied:
well, the iso7816 driver is quite successful in offering a generic and extensible implementation. For ias/ecc the same should be possible. Unfortunately, I think it's already too late...

martinpaljak commented:
> Please use tabs for indenting. Looks quite OK, nothing serious.
It has been some time since I wrote C so my setup has changed. Using VS Code with clang-format seems pretty neat. Maybe we can add necessary tooling/config into the source tree so that code formatting would be automated with a tool (like clang-format) ? I'll keep using the default formatting of VScode+clang-format plugin until the PR is non-draft or until I figure out how to configure clang-format in a suitable way.

frankmorgner commented:
> I'll keep using the default formatting of VScode+clang-format plugin until the PR is non-draft or until I figure out how to configure clang-format in a suitable way.

The way of achieving unified code style in OpenSC or recommendations thereof is still a todo. Sure, you can postpone a (somewhat) unified style for later.

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
+	LOG_TEST_RET(card->ctx, r, "read docnr failed");
+	const unsigned char *tag = sc_asn1_find_tag(card->ctx, buff, r, 0x04, &taglen);
```
(size_t) r

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+	for (i = 0; i < 2; i++) {
+		const char *esteid_cert_names[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *esteid_cert_paths[2] = {"3f00:adf1:3401", "3f00:adf2:341f"};
+		const int esteid_cert_ids[2] = {1, 2};
```
const u8 esteid_cert_ids

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+		static const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
+		static const int esteid_pin_min[3] = {4, 5, 8};
+		static const int esteid_pin_ref[3] = {0x01, 0x85, 0x02};
+		static const int esteid_pin_authid[3] = {1, 2, 3};
```
static const u8 esteid_pin_authid

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+		static const int esteid_pin_authid[3] = {1, 2, 3};
+		static const char *esteid_pin_path[3] = {"3F00", "3F00ADF2", "3F00"};
+
+		static const int esteid_pin_flags[3] = {
```
static const unsigned int esteid_pin_flags

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		static const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
+		static const int esteid_pin_min[3] = {4, 5, 8};
```
static const size_t esteid_pin_min

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+
+	/* add private keys */
+	for (i = 0; i < 2; i++) {
+		int prkey_pin[2] = {1, 2};
```
const u8 prkey_pin

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,223 @@
+
+		const char *prkey_name[2] = {"Isikutuvastus", "Allkirjastamine"};
+		const char *prkey_path[2] = {"3F00:ADF1", "3F00:ADF2"};
+		const int prkey_usage[2] = {SC_PKCS15_PRKEY_USAGE_SIGN | SC_PKCS15_PRKEY_USAGE_DERIVE,
```
const unsigned int prkey_usage

metsma commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,347 @@
+
+static int esteid_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out) {
+	unsigned char pathbuf[SC_MAX_PATH_SIZE], *path = pathbuf;
+	int pathlen;
```
size_t pathlen;

metsma commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,347 @@
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "PSO CDS/INTERNAL AUTHENTICATE failed");
+
+	LOG_FUNC_RETURN(card->ctx, apdu.resplen);
```
(int) apdu.resplen

metsma commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,347 @@
+static int esteid_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t count, unsigned long flags) {
+	int r;
+	LOG_FUNC_CALLED(card->ctx);
+	void *saved = card->ops->check_sw;
```
typedef int (*check_sw)(struct sc_card *card,unsigned int sw1,unsigned int sw2);
	check_sw saved = card->ops->check_sw;

martinpaljak replied:
Correct me if I'm wrong, but handling a void pointer in this context should be sufficient effectively (a pointer is a pointer no matter where it points)

metsma replied:
https://stackoverflow.com/questions/36645660/why-cant-i-cast-a-function-pointer-to-void

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+	while (pathlen >= 2) {
+		if (memcmp(path, "\x3F\x00", 2) == 0) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "MF select failed");
+		} else if (pathlen >= 2 && path[0] == 0xAD) {
```
pathlen will be at least 2, due to the `while` condition

martinpaljak replied:
Indeed

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+	struct esteid_priv_data *priv = DRVDATA(card);
+	struct sc_security_env *env = NULL;
+	struct sc_apdu apdu;
+	u8 sbuf[SIGNATURE_PAYLOAD_SIZE] = {0};
```
Either use `= {0};` or `memset(sbuf, 0x00, SIGNATURE_PAYLOAD_SIZE - datalen);`, but don't use both. The former should work starting with C99.

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+
+	switch (env->key_ref[0]) {
+	case 1: /* authentication key */
+		sc_format_apdu_ex(card, &apdu, 0x88, 0, 0, sbuf, datalen, out, MIN(256, outlen));
```
Do you want to use `SC_MAX_APDU_RESP_SIZE` instead of `256`?

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+
+	get_pin_info[6] = pin_reference & 0x0F; // mask out local/global
+	// XXX: handling of default Le of 0x00 could be easier
+	sc_format_apdu_ex(card, &apdu, 0xCB, 0x3F, 0xFF, get_pin_info, sizeof(get_pin_info), apdu_resp, MIN(256, sizeof(apdu_resp)));
```
Use `SC_MAX_APDU_RESP_SIZE` for declaring `apdu_resp` instead of `SC_MAX_APDU_BUFFER_SIZE` and you're good to specify the length with `sizeof(apdu_resp)` only

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+
+	// XXX: sc_asn1_find_tag with the following payload (to get to tag 0x9B):
+	// https://lapo.it/asn1js/#cB6_gQEaoBiaAQObAQOhEIwG8wAAc0MAnAbzAABzQwA
+	return apdu_resp[13];
```
Yes, parsing the content is the way to go! At the very least, you should check whether you've received at least 14 bytes.

frankmorgner replied:
cast to int?

martinpaljak replied:
Check

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+		tmp.pin_reference = 0x02; // hardcoded, ugly
+		tmp.pin2.len = 0;
+		r = iso_ops->pin_cmd(card, &tmp, tries_left);
+		sc_mem_clear(&tmp, sizeof(tmp));
```
no need to clear the memory, because the actual PINs (`tmp.pin1.data`/`tmp.pin2.data`) are only pointers.

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+	} else if (data->cmd == SC_PIN_CMD_UNBLOCK) {
+		// Verify PUK, then issue UNBLOCK
+		// VERIFY
+		memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
```
why don't you use `tmp = *data`?

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,327 @@
+		// VERIFY
+		memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
+		tmp.cmd = SC_PIN_CMD_VERIFY;
+		tmp.pin_reference = 0x02; // hardcoded, ugly
```
use a human readable define instead of `0x02`.

frankmorgner commented on src/libopensc/iso7816.c:
```diff
@@ -152,10 +152,8 @@ iso7816_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t coun
-		LOG_FUNC_RETURN(ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
 
-	r =  sc_check_sw(card, apdu.sw1, apdu.sw2);
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
```
This change looks OK.

With this applied, do you still need the workaround with `esteid_check_sw`?

martinpaljak replied:
Yes.

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,220 @@
+
+static int sc_pkcs15emu_esteid2018_init(sc_pkcs15_card_t *p15card) {
+	sc_card_t *card = p15card->card;
+	u8 buff[128];
```
looking at your code below, 11 bytes should be enough for buff

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,220 @@
+	/* Read docnr */
+	sc_format_path("3F00D003", &tmppath);
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &tmppath, NULL), "SELECT docnr");
+	r = sc_read_binary(card, 0, buff, 11, 0);
```
sizeof buff instead of 11?

frankmorgner commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,220 @@
+
+	/* add pins */
+	for (i = 0; i < 3; i++) {
+		const char *esteid_pin_names[3] = {"PIN1", "PIN2", "PUK"};
```
aren't there better names for PIN1 and PIN2?

metsma replied:
This is how they are labeled in PIN envelope when they are handed to citizens

frankmorgner replied:
Then the *real* fix is needed elsewhere...

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+    // VERIFY
+    memcpy(&tmp, data, sizeof(struct sc_pin_cmd_data));
+    tmp.cmd = SC_PIN_CMD_VERIFY;
+    tmp.pin_reference = 0x02; // hardcoded, ugly
```
What do you mean?

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,411 @@
+    LOG_TEST_RET(card->ctx, r, "VERIFY during unblock failed");
+
+    if (data->pin_reference == 0x85) {
+      LOG_TEST_RET(card->ctx, esteid_select_file(card, &adf2, NULL),
```
You mean the caller?

frankmorgner commented:
please also look at the two unresolved comments above

metsma commented on src/libopensc/pkcs15-esteid2018.c:
```diff
@@ -0,0 +1,220 @@
+			continue;
+
+		sc_pkcs15_cert_t *cert = NULL;
+		r = sc_pkcs15_read_certificate(p15card, &cert_info, &cert);
```
Fails to read Test card certificate. Reads only 1024 bytes.
Seems like the size is hardcoded here
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15.c#L2355

martinpaljak replied:
Indeed. fixed this with latest push to this branch. But the right thing to do would be to remove that fixed 1k buffer from core so that the file could be read "until fully read" irrelevant of size.

frankmorgner commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,330 @@
+
+	// XXX: sc_asn1_find_tag with the following payload (to get to tag 0x9B):
+	// https://lapo.it/asn1js/#cB6_gQEaoBiaAQObAQOhEIwG8wAAc0MAnAbzAABzQwA
+	return (int)apdu_resp[13];
```
need to check response size

martinpaljak commented:
> See comment above. There is also [this remark](https://github.com/OpenSC/OpenSC/pull/1635/files#r268146297). Please exclude `.clang-format` from this PR.
> 
> The rest looks good.

Addressed both, removed clang-format, rebased on master.

martinpaljak commented on src/libopensc/card-esteid2018.c:
```diff
@@ -0,0 +1,333 @@
+
+static int esteid_match_card(sc_card_t *card) {
+	int i = 0;
+	int r = 0;
```
Indeed, reduced in last push, that rebases it against master and reduces number of "review updates" commits.
